### PR TITLE
Teach clone to recurse into submodules

### PR DESCRIPTION
**Pull request by jamill**
This is the basic outline to enable cloning with the option to recursively clone submodules. I wanted to put this up to get feedback on how the progress is reported through callbacks.

**Notes:**
This adds callbacks that are called before and after cloning a repository. The user has an opportunity to cancel the clone before starting a clone operation in a repository.

### CODE REVIEW

**Comment by jamill** [comment]
Not the biggest fan of the `CurrentRepository` prefix on the repository callbacks and the enum, but using it until I come up with a better name.

**Comment by jamill** [comment]
I think this could supersede #640

File: LibGit2Sharp/RepositoryOperationContext.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        /// <summary>
+        /// Constructor suitable for use on the sub repositories.
+        /// </summary>
+        /// <param name="repositoryPath"></param>
+        /// <param name="parentRepositoryPath"></param>
+        /// <param name="submoduleName"></param>
+        /// <param name="recursionDepth"></param>
+        public RepositoryOperationContext(string repositoryPath, string parentRepositoryPath, string submoduleName, int recursionDepth)
```
Comment (nulltoken) [inline LibGit2Sharp/RepositoryOperationContext.cs]: Can we make this internal?

File: LibGit2Sharp/RepositoryOperationContext.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    /// Class to convey information about the repository that is being operated on
+    /// for operations that can recurse into submodules.
+    /// </summary>
+    public class RepositoryOperationContext
+    {
+        /// <summary>
+        /// Needed for mocking.
+        /// </summary>
+        protected RepositoryOperationContext()
+        {
```
Comment (nulltoken) [inline LibGit2Sharp/RepositoryOperationContext.cs]: Could you please put the opening and closing braces on the same line?

File: LibGit2Sharp.Tests/CloneFixture.cs [hunk 3]
```diff
@@ -242,5 +244,239 @@ public void CloningWithoutUrlThrows()
...
+                Assert.Equal(kvp.Value.RemoteUrl, entry.RemoteUrl);
+                Assert.Equal(kvp.Value.RecursionDepth, entry.RecursionDepth);
+                Assert.Equal(kvp.Value.StartingWorkInRepositoryCalled, entry.StartingWorkInRepositoryCalled);
+                Assert.Equal(kvp.Value.FinishedWorkInRepositoryCalled, entry.FinishedWorkInRepositoryCalled);
+                Assert.Equal(kvp.Value.CheckoutProgressCalled, entry.CheckoutProgressCalled);
+                Assert.Equal(kvp.Value.RemoteRefUpdateCalled, entry.RemoteRefUpdateCalled);
+            }
+
+            // submodule is initialized
+            // To Verify: submodule head commit
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/CloneFixture.cs]: Is that a reminder?

File: LibGit2Sharp/Repository.cs [hunk 4]
```diff
@@ -574,13 +594,151 @@ public static string Discover(string startingPath)
...
+        /// <param name="options">Options controlling clone behavior.</param>
+        /// <param name="repoPath">Path of the parent repository.</param>
+        /// <param name="recursionDepth">The current depth of the recursion.</param>
+        private static bool RecursivelyCloneSubmodules(CloneOptions options, string repoPath, int recursionDepth)
+        {
+            bool continueOperation = true;
+
+            if (options.RecurseSubmodules)
+            {
+                    List<string> submodules = new List<string>();
```
Comment (nulltoken) [inline LibGit2Sharp/Repository.cs]: Is there some indentation issue here?

**Comment by jamill** [comment]
@nulltoken How do you feel about this? I added a new exception to indicate when an error was encountered while recursively cloning submodules.

**Comment by nulltoken** [comment]
@jamill Something weird is going on. Running the CloneFixture trashes the test runners (repo'd with TestDiven.net and Resharper). I've repro'd this as well by removing all the fixtures from the project (by `CloneFixture`) and running `.\build.libgit2sharp.cmd` in a Powershell session.

FWIW, the call stack looks suspiciously similar to https://github.com/libgit2/libgit2sharp/pull/895#issuecomment-80411460

image

/cc @ethomson

**Comment by jamill** [comment]
Unfortunate - I hadn't seen this issue before. Are we seeing this issue even without the new clone changes? Will try to take a look, but probably won't be right away.

Sent from my Windows Phone
________________________________

**Comment by nulltoken** [comment]
@jamill I *think* it's unrelated to the clone work as the Xunit20 branch fails as well.

**Comment by nulltoken** [comment]
@jamill Ok to squash and merge?

**Comment by nulltoken** [comment]
Published as NuGet pre-release package `LibGit2Sharp.0.22.0-pre20150409142943`

**Pull request merged**